Q1) [C, Python] The probability of rain on a given calendar day in Vancouver is p[i], where i is the day's index. For
example, p[0] is the probability of rain on January 1st, and p[10] is the probability of precipitation on January 11th.

Assume the year has 365 days (i.e., p has 365 elements). What is the chance it rains more than n (e.g., 100) days in Vancouver?
Write a function that accepts p (probabilities of rain on a given calendar day) and n as input arguments and returns the
possibility of raining at least n days.


def prob_rain_more_than_n(p: Sequence[float], n: int) -> float:
pass

Assumption made: Using the Monte Carlo simulation for this problem as it handles the complexity of varying daily rain probabilities well. Each day has its own chance of rain, so using straightforward analytical methods like the binomial distribution isn’t an ideal approach.

Reasons:
Firstly, Monte Carlo simulation is great because it simulates the entire year thousands of times, and shows how often we end up with more than a certain number of rainy days. 
Secondly, it's flexible, in order to tweak the probabilities or the number of days, I can do that without reworking the whole approach. 
Finally, the more simulations I run, the closer I get to the true probability, which helps in making a more accurate guess.

In [43]:
from typing import Sequence
import random


def prob_rain_more_than_n(p: Sequence[float], n: int) -> float:
    """
    Args:
    p: List of 365 daily rain probabilities.
    n: Threshold for rainy days
    num_sim: Number of simulations (10,000).

    Returns:
    Probability as a float (0 to 1). Multiplying by 100 for percentage.

    """
    num_sim = 10000
    count_greater_n = 0

    #monte Carlo Sim
    for _ in range(num_sim):
        rainy_days = 0

        for day in range(365):
            if random.random() < p[day]:
                rainy_days += 1
        
        if rainy_days > n:
            count_greater_n += 1
    
    return count_greater_n / num_sim



In [51]:
#Test case
p = [0.3] * 365  # Test where each day has a 30% chance of rain
n = 100
probability = prob_rain_more_than_n(p, n)
print(f"{probability* 100:.2f}% is the possibility of it raining atleast more than {n} days for the year")

84.47% is the possibility of it raining atleast more than 100 days for the year


Q2)  A phoneme is a sound unit (similar to a character for text). We have an extensive pronunciation
dictionary (think millions of words). Below is a snippet:

ABACUS AE B AH K AH S

BOOK B UH K

THEIR DH EH R

THERE DH EH R

TOMATO T AH M AA T OW

TOMATO T AH M EY T OW

Given a sequence of phonemes as input (e.g. ["DH", "EH", "R", "DH", "EH", "R"]), find all the combinations of the words that
can produce this sequence (e.g. [["THEIR", "THEIR"], ["THEIR", "THERE"], ["THERE", "THEIR"], ["THERE", "THERE"]]).

You can preprocess the dictionary into a different data structure if needed.

def find_word_combos_with_pronunciation(phonemes: Sequence[str]) -> Sequence[Sequence[str]]:
pass

Assumptions made: Preprocessing will lead to efficiency. 
Preprocessing/organizing the data will help find words that match the sequence of phonemes quickly. 

Part # 1:
Process: The pronunciation dictionary is turned into a map where each key is a tuple of phonemes, and the value is a list of words that match those phonemes. 

Reason: Looking up words to match parts of the phoneme sequence can be done in an instant.

Part # 2:
Process: Backtracking is like exploring a maze where you try different paths to see which one leads to the exit. 
In this case, the "maze" is our sequence of phonemes, and the "exit" is finding all the ways we can split this sequence into valid words. 

The process begins at the start of the phoneme sequence, attempting to match the longest possible prefix to a word. If a match is found, the remaining sequence is checked recursively for additional matches. If a path fails, backtracking occurs and tries an alternative path. This systematic approach ensures that all valid word combinations are considered. A preprocessed dictionary enhances efficiency, making the process significantly faster.

In [76]:
from typing import Sequence, List, Dict, Tuple

def preprocess_dictionary(pronunciation_dict: List[Tuple[str, List[str]]]):
    """
    Preprocess the pronunciation dictionary into a mapping from phoneme sequences to words.
    """
    phoneme_to_words = {}
    for word, phonemes in pronunciation_dict:
        phoneme_tuple = tuple(phonemes)
        if phoneme_tuple not in phoneme_to_words:
            phoneme_to_words[phoneme_tuple] = []
        phoneme_to_words[phoneme_tuple].append(word)
    return phoneme_to_words

def find_word_combos_with_pronunciation(phonemes: Sequence[str], phoneme_to_words: Dict[Tuple[str, ...], List[str]]) -> Sequence[Sequence[str]]:
    """
    Find all combinations of words that can produce the given sequence of phonemes.
    """


    def backtrack(start: int) -> List[List[str]]:
        if start == len(phonemes):
            return [[]]  # Base case: return a list with an empty list

        results = []
        for end in range(start + 1, len(phonemes) + 1):
            current_phoneme_tuple = tuple(phonemes[start:end])
            if current_phoneme_tuple in phoneme_to_words:
                words = phoneme_to_words[current_phoneme_tuple]
                for word in words:
                    for rest in backtrack(end):
                        results.append([word] + rest)
        return results

    return backtrack(0)

In [77]:
# Test Case 1:
pronunciation_dict = [
    ("ABACUS", ["AE", "B", "AH", "K", "AH", "S"]),
    ("BOOK", ["B", "UH", "K"]),
    ("THEIR", ["DH", "EH", "R"]),
    ("THERE", ["DH", "EH", "R"]),
    ("TOMATO", ["T", "AH", "M", "AA", "T", "OW"]),
    ("TOMATO", ["T", "AH", "M", "EY", "T", "OW"])
]

"""
Assumption Made: the words produced as an output are always supposed to correspond to the 
order of  the phonemes sequence list/input.

"""

phoneme_to_words = preprocess_dictionary(pronunciation_dict)
phonemes = ["DH", "EH", "R", "DH", "EH", "R"]
combinations = find_word_combos_with_pronunciation(phonemes, phoneme_to_words)
print(combinations)

[['THEIR', 'THEIR'], ['THEIR', 'THERE'], ['THERE', 'THEIR'], ['THERE', 'THERE']]


In [78]:
# Test Case 2:
pronunciation_dict = [
    ("APPLE", ["AE", "P", "AH", "L"]),
    ("PEAR", ["P", "EH", "R"]),
    ("BEAR", ["B", "EH", "R"]),
    ("PAIR", ["P", "EH", "R"]),
    ("PEEL", ["P", "IY", "L"]),
    ("BELL", ["B", "EH", "L"])
]

phoneme_to_words = preprocess_dictionary(pronunciation_dict)
phonemes = ["P", "EH", "R", "B", "EH", "L"]
phonemes_2 = ["P", "EH", "R","P", "EH", "R"] #try_2
combinations = find_word_combos_with_pronunciation(phonemes, phoneme_to_words)
print(combinations)

[['PEAR', 'BELL'], ['PAIR', 'BELL']]


Q3) Implement CTC as this [paper](https://dl.acm.org/doi/abs/10.1145/1143844.1143891) describes. Your implementation should support both forward and
backward propagation operations.

Assumptions

In implementing the Connectionist Temporal Classification (CTC) algorithm, we assume that the input consists of sequences of log probabilities, which are used to maintain numerical stability during calculations. A special blank token is employed to facilitate the alignment of input sequences with target label sequences, allowing for the insertion of gaps between labels. The target label sequence is shorter than or equal to the input sequence length and does not initially contain blanks.


Insert Blanks: The insert_blanks function modifies the target label sequence by inserting blank tokens between each label and at the start and end. 
This prepares the sequence for the forward and backward passes.

In [67]:
import numpy as np
"""
Args:

labels: List of target labels (integers).
blank: Index for the blank label (default is 0).
Returns:

list: A new list with blanks inserted.

"""
def insert_blanks(labels, blank=0):
    new_labels = [blank]
    for label in labels:
        new_labels.extend([label, blank])
    return new_labels

Forward propagation:

The forward_pass function computes the forward variables using dynamic programming. This step calculates the probability of reaching each state at every time step.

In [69]:
def forward_pass(log_probs, labels, blank=0):
    """
    Compute the forward variables 'alpha' for CTC using dp.
    
    Args:
        log_probs: Log probabilities of shape (T, num_classes).
        labels: Target label sequence.
        blank: Index for the blank label.
    
    Returns:
        np.ndarray: Forward variables.
    """
    
    T, _ = log_probs.shape
    ext_labels = insert_blanks(labels, blank)
    S = len(ext_labels)
    
    alpha = np.full((T, S), -np.inf)
    alpha[0, 0] = log_probs[0, blank]
    if S > 1:
        alpha[0, 1] = log_probs[0, ext_labels[1]]
    
    for t in range(1, T):
        for s in range(S):
            candidates = [alpha[t-1, s]]
            if s > 0:
                candidates.append(alpha[t-1, s-1])
            if s > 1 and ext_labels[s] != blank and ext_labels[s] != ext_labels[s-2]:
                candidates.append(alpha[t-1, s-2])
            max_val = np.max(candidates)
            if np.isfinite(max_val):
                alpha[t, s] = max_val + np.log(np.sum(np.exp(np.array(candidates) - max_val)))
            alpha[t, s] += log_probs[t, ext_labels[s]]
    
    return alpha

Backward Propagation: 

The backward_pass function computes the backward variables, representing the probability of completing the sequence from each state, also using dynamic programming.

In [70]:
def backward_pass(log_probs, labels, blank=0):
    """
    Compute the backward variables 'beta' for CTC using dp.
    
    Args:
        log_probs: Log probabilities of shape (T, num_classes).
        labels: Target label sequence.
        blank: Index for the blank label.
    
    Returns:
        np.ndarray: Backward variables.
    """
    T, _ = log_probs.shape
    ext_labels = insert_blanks(labels, blank)
    S = len(ext_labels)
    
    beta = np.full((T, S), -np.inf)
    beta[T-1, S-1] = log_probs[T-1, blank]
    if S > 1:
        beta[T-1, S-2] = log_probs[T-1, ext_labels[S-2]]
    
    for t in range(T-2, -1, -1):
        for s in range(S):
            candidates = [beta[t+1, s]]
            if s < S-1:
                candidates.append(beta[t+1, s+1])
            if s < S-2 and ext_labels[s] != blank and ext_labels[s] != ext_labels[s+2]:
                candidates.append(beta[t+1, s+2])
            max_val = np.max(candidates)
            if np.isfinite(max_val):
                beta[t, s] = max_val + np.log(np.sum(np.exp(np.array(candidates) - max_val)))
            beta[t, s] += log_probs[t, ext_labels[s]]
    
    return beta

CTC Loss and Gradient: The ctc_loss_and_grad function combines the forward and backward results to compute the total probability of the target sequence. 

The loss is the negative log of this probability, and the gradient is calculated for use in backpropagation.

In [71]:
def ctc_loss_and_grad(log_probs, labels, blank=0):

    """
    Computes the CTC loss and gradient with respect to the network outputs.

    Args:

    log_probs: Log probabilities of shape (T, num_classes).
    labels: Target label sequence.
    blank: Index for the blank label.
    
    Returns:

    float: CTC loss.
    np.ndarray: Gradient of the loss with respect to log_probs.

    """
    
    T, num_classes = log_probs.shape
    alpha = forward_pass(log_probs, labels, blank)
    beta = backward_pass(log_probs, labels, blank)
    ext_labels = insert_blanks(labels, blank)
    S = len(ext_labels)
    
    log_prob1 = alpha[T-1, S-1]
    log_prob2 = alpha[T-1, S-2] if S > 1 else -np.inf
    total_log_prob = np.logaddexp(log_prob1, log_prob2)
    
    if not np.isfinite(total_log_prob):
        print("Warning: total_log_prob is not finite.")
    
    loss = -total_log_prob
    
    grad = np.zeros_like(log_probs)
    for t in range(T):
        for s in range(S):
            k = ext_labels[s]
            grad[t, k] -= np.exp(alpha[t, s] + beta[t, s] - total_log_prob)
    
    return loss, grad

yielding the CTC loss and gradient necessary for training a neural network.

In [72]:
if __name__ == '__main__':
    T = 50  # Number of time steps
    num_classes = 5  # Number of classes including the blank
    np.random.seed(42)
    logits = np.random.randn(T, num_classes)
    log_probs = np.log(np.exp(logits) / np.sum(np.exp(logits), axis=1, keepdims=True))
    
    target = [1, 2, 3]  # Example target label sequence
    
    loss, grad = ctc_loss_and_grad(log_probs, target, blank=0)
    
    print("CTC Loss:", loss)
    print("CTC Gradient shape:", grad.shape)

CTC Loss: 69.05861912394417
CTC Gradient shape: (50, 5)


CTC Loss: The value 69.05861912394417 represents the negative log likelihood of the target label sequence given the input sequence. 
In the context of training a neural network, this loss value would be used to update the model's weights to improve its predictions. 
A lower loss indicates a better alignment between the predicted and target sequences.

Gradient Shape: The gradient shape (50, 5) indicates that the gradient has been computed for each time step (50 in total) and each class (5 classes, including the blank). 
This gradient is used to adjust the network's weights during backpropagation to minimize the CTC loss.

- Numerical Stability: Uses log-space calculations to prevent underflow or overflow, ensuring stability when working with probabilities.
- Dynamic Programming: Utilizes dynamic programming in both forward and backward passes to efficiently compute alignment probabilities between input and target sequences.
- Gradient Calculation: Computes gradients for each time step and class, enabling backpropagation to update network weights.
- Integration: Designed for seamless integration into a neural network training pipeline, optimizing model parameters using CTC loss and gradients.

the CTC handles unsegmented sequence data, such as speech, by aligning input sequences with target labels without requiring presegmented data.